### ISSUE TITLE
Trigram indexes

### ISSUE THREAD

**Issue by deontologician** [rethinkdb/rethinkdb#5351]
The benefit of having an explicit trigram index vs. convenience functions to create generic multi-indexes is that you can build things like regex search on top if you know the underlying index has trigram invariants:

https://swtch.com/~rsc/regexp/regexp4.html

The linked article is about RE2 style regexes (i.e. no backtracking etc), and it would allow us to do things like `table('files').match('bleh.*\.jpg', index='filename_trigram')`

This is related to #2601 because full text search might be built off trigram indexes as well, though this is a more narrowly scoped.

**Comment by danielmewes** [comment]
Oh that's cool.
A trigram index itself could be created using our existing indexes with a custom index function right?
So the interesting part it seems it implementing `match` on such an index.

**Comment by marshall007** [comment]
This is essentially a special case of #4226. Supporting a (limited) indexed `.match()` on top of that would be wonderful.

**Comment by marshall007** [comment]
Also, would be nice to have some syntactic sugar allowing you to perform an indexed `.match()` across multiple indices in a single query. For example:

```js
r.table('files').match('foo.*bar', index=[ 'filename_trigram', 'file_contents_trigram' ])

// roughly equivalent to:

r.union(
  r.table('files').match('foo.*bar', index='filename_trigram'),
  r.table('files').match('foo.*bar', index='file_contents_trigram')
).distinct()
```

The main reason I think this is preferable over the `union` workaround is because it might allow for the possibility of an efficient [`minimum_should_match`](https://www.elastic.co/guide/en/elasticsearch/reference/2.2/query-dsl-multi-match-query.html#operator-min) and other equivalent options from Elasticsearch in the future.

**Comment by deontologician** [comment]
@danielmewes yeah, I'm thinking this could be done using the same index machinery, only maybe with a flag that indicates it was constructed as a trigram and can be used for certain queries.

Also thanks for finding #4226 @marshall007 I had forgotten about that one. Technically I guess ngrams would be more general than a trigram index, but that issue seems narrowly focused on prefixes only

**Comment by wrathagom** [comment]
Just food for thought if you've never used elastic search or postgres' ngram functionality. Might help lay some foudation or see various use cases. I too would be interested in this.

http://www.postgresql.org/docs/9.1/static/pgtrgm.html
https://www.elastic.co/guide/en/elasticsearch/guide/current/_ngrams_for_partial_matching.html
